server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesUpdateProcessor.java

```diff
@@ -85,27 +77,20 @@ public SetMessagesResponse process(SetMessagesRequest request,  MailboxSession m
     private void update(MessageId messageId, UpdateMessagePatch updateMessagePatch, MailboxSession mailboxSession,
                         SetMessagesResponse.Builder builder) {
         try {
-            MessageMapper messageMapper = mailboxSessionMapperFactory.createMessageMapper(mailboxSession);
-            Mailbox mailbox = mailboxMapperFactory.getMailboxMapper(mailboxSession)
-                    .findMailboxByPath(messageId.getMailboxPath());
-            Iterator<MailboxMessage> mailboxMessage = messageMapper.findInMailbox(
-                    mailbox, MessageRange.one(messageId.getUid()), MessageMapper.FetchType.Metadata, LIMIT_BY_ONE);
-            MailboxMessage messageWithUpdatedFlags = applyMessagePatch(messageId, mailboxMessage.next(), updateMessagePatch, builder);
-            savePatchedMessage(mailbox, messageId, messageWithUpdatedFlags, messageMapper);
+            MessageManager messageManager = mailboxManager.getMailbox(messageId.getMailboxPath(), mailboxSession);
+            MessageResultIterator message = messageManager.getMessages(messageId.getUidAsRange(), FetchGroupImpl.MINIMAL, mailboxSession);
+            updateFlags(messageId, updateMessagePatch, mailboxSession, messageManager, message.next());
+            builder.updated(ImmutableList.of(messageId));
         } catch (NoSuchElementException e) {
             addMessageIdNotFoundToResponse(messageId, builder);
         } catch (MailboxException e) {
             handleMessageUpdateException(messageId, builder, e);
         }
     }
 
-    private boolean savePatchedMessage(Mailbox mailbox, MessageId messageId,
-                                       MailboxMessage message,
-                                       MessageMapper messageMapper) throws MailboxException {
-        return messageMapper.updateFlags(mailbox, new FlagsUpdateCalculator(message.createFlags(),
-                        MessageManager.FlagsUpdateMode.REPLACE),
-                MessageRange.one(messageId.getUid()))
-                .hasNext();
+    private void updateFlags(MessageId messageId, UpdateMessagePatch updateMessagePatch, MailboxSession mailboxSession, MessageManager messageManager, MessageResult messageResult) throws MailboxException {
+        Flags newState = updateMessagePatch.applyToState(messageResult.getFlags());
```

Comment:
I agree but we should open a ticket for that.
Keep in mind 3.0 should be release soon and jmap is just experimental, we should not waste time on it right now.